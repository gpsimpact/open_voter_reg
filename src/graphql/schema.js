import glob from 'glob';
import fs from 'fs';
import path from 'path';
import { makeExecutableSchema } from 'graphql-tools';

process.on('warning', e => console.warn(e.stack));

const typeDefs = [];

// loop over types and import the contents
glob.sync('**/*.type.gql', { cwd: __dirname }).forEach(filename => {
  const filePath = path.join(__dirname, filename);
  const fileContent = fs.readFileSync(filePath, { encoding: 'utf8' });
  typeDefs.push(fileContent);
});

import resolvers from './resolvers';

// create a schema
const schema = makeExecutableSchema({ typeDefs, resolvers });

// add mocks to the schema, preserving the existing resolvers (none for the time beeing)
// import mocks from './mocks';
// addMockFunctionsToSchema({ schema, mocks, preserveResolvers: true });

export default schema;
