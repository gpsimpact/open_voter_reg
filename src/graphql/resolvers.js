import QueryResolver from './Query/Query.resolvers';
import MutationResolver from './Mutation/Mutation.resolvers';

export default {
  ...QueryResolver,
  ...MutationResolver,
};
