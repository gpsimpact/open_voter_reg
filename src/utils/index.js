import faker from 'faker';
import { includes } from 'lodash';
import { InsufficientPermissionsError } from '../errors';

export function mapTo(keys, keyFn) {
  return rows => {
    const group = new Map(keys.map(key => [key, null]));
    rows.forEach(row => group.set(keyFn(row), row));
    return Array.from(group.values());
  };
}

export function mapToMany(keys, keyFn) {
  return rows => {
    const group = new Map(keys.map(key => [key, []]));
    rows.forEach(row => (group.get(keyFn(row)) || []).push(row));
    return Array.from(group.values());
  };
}

export function mapToValues(keys, keyFn, valueFn) {
  return rows => {
    const group = new Map(keys.map(key => [key, null]));
    rows.forEach(row => group.set(keyFn(row), valueFn(row)));
    return Array.from(group.values());
  };
}

export const generateFakeUsers = (num, seed) => {
  faker.seed(seed); // ensures consistent result
  const fakeUsers = [];
  Array(num)
    .fill()
    .map((_, i) => {
      fakeUsers.push({
        first_name: faker.name.firstName(),
        last_name: faker.name.lastName(),
        email: faker.internet.email(),
        password: faker.internet.password(),
      });
    });
  return fakeUsers;
};

export const hasPermission = (viewer, service, requiredPermission, throwOnFail = false) => {
  if (!includes(viewer.permissions[service], requiredPermission)) {
    if (throwOnFail) {
      throw new InsufficientPermissionsError();
    }
    return false;
  }
  return true;
};
