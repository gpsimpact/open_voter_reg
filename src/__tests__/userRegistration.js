import { graphql } from 'graphql';
import db from '../db';
import schema from '../graphql/schema';
// import faker from 'faker';
import MakeContext from '../Context';
import { find } from 'lodash';
import bcrypt from 'bcrypt';
import { generateFakeUsers } from '../utils';

beforeAll(async () => await db.migrate.latest({ directory: './src/db/migrations' }));
beforeEach(async () => await Promise.all([db.raw('TRUNCATE TABLE users CASCADE ')]));
afterAll(async () => await db.destroy());

describe('User Registration', () => {
  // beforeEach(async () => await initializeTestingDB());

  test('Registering new user with existing email should return error', async () => {
    //// setup
    const users = generateFakeUsers(1, 123);
    await db('users').insert(users[0]);

    //// test
    const query = `
      mutation {
          registerUser(
            user:{
              first_name: "${users[0].first_name}", 
              last_name: "${users[0].last_name}", 
              email: "${users[0].email}",
              password: "${users[0].password}",
            }
          ) {
            first_name
            last_name
            email
          }
        }
    `;

    const rootValue = {};
    const context = new MakeContext({ user: null });
    const result = await graphql(schema, query, rootValue, context);
    expect(result).toMatchSnapshot();
    // const { data } = result;
    expect(result.errors.length).toBeGreaterThanOrEqual(1);
    expect(
      find(result.errors, { message: 'This email address is already associated with an account.' })
    ).not.toBeUndefined();
  });

  test('password should be correctly encyrpted', async () => {
    const users = generateFakeUsers(1, 123);
    const query = `
        mutation {
            registerUser(
              user:{
                first_name: "${users[0].first_name}", 
                last_name: "${users[0].last_name}", 
                email: "${users[0].email}",
                password: "${users[0].password}",
              }
            ) {
              first_name
              last_name
              email
            }
          }
      `;

    const rootValue = {};
    const context = new MakeContext({ user: null });
    await graphql(schema, query, rootValue, context);
    // Grab record from database
    const dbUserRecord = await db('users')
      .first()
      .where({ email: users[0].email });
    expect(dbUserRecord.password).not.toBeUndefined();
    expect(bcrypt.compareSync(users[0].password, dbUserRecord.password)).toBe(true);
  });

  test('new registrant emailVerified should default to false', async () => {
    const users = generateFakeUsers(1, 123);
    const query = `
        mutation {
            registerUser(
              user:{
                first_name: "${users[0].first_name}", 
                last_name: "${users[0].last_name}", 
                email: "${users[0].email}",
                password: "${users[0].password}",
              }
            ) {
              first_name
              last_name
              email
              email_verified
            }
          }
      `;

    const rootValue = {};
    const context = new MakeContext({ user: null });
    const result = await graphql(schema, query, rootValue, context);
    // console.log(JSON.stringify(result, null, '\t'));
    // expect(result.data).toMatchSnapshot();
    expect(result.data.registerUser.email_verified).toBe(false);
    // Grab record from database
    const dbUserRecord = await db('users')
      .first()
      .where({ email: users[0].email });
    expect(dbUserRecord.email_verified).toBe(false);
  });
});
