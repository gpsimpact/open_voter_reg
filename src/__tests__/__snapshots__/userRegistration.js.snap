// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Registration Registering new user with existing email should return error 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: This email address is already associated with an account.],
  ],
}
`;
